File: 백준/Silver/9012. 괄호/괄호.cs
int t = int.Parse(Console.ReadLine());


for (int i = 0; i < t; i++)
{
    Stack<char> stack = new Stack<char>();
    string input = Console.ReadLine();

    for (int j = 0; j < input.Length; j++)
    {
        if (stack.Count == 0)
        {
            stack.Push(input[j]);
        }
        else
        {
            if (stack.Peek() == '(' && input[j] == ')')
            {
                stack.Pop();
            }
            else
            {
                stack.Push(input[j]);
            }
        }
    }

    if (stack.Count == 0)
    {
        Console.WriteLine("YES");
    }
    else
    {
        Console.WriteLine("NO");
    }
}

File: 백준/Silver/9012. 괄호/괄호.cc
#include <iostream>
#include <stack>

using namespace std;

bool CheckStr(string str)
{
  stack<char> par;

  for(int i = 0; i<str.size(); i++)
  {
    if(str[i] == '(')
      par.push(str[i]);
    else
      {
        if(!par.empty())
          par.pop();
        else
          return false;
      }
  }

  return par.empty();
}

int main()
{
  int t;

  cin >> t;

  for(int i = 0; i<t; i++)
  {
    string tmp;
    cin >> tmp;

    if(CheckStr(tmp))
      cout << "YES" << endl;
    else
      cout << "NO" << endl;
  }


  return 0;
}
